numpy/core/defchararray.py

```diff
@@ -51,21 +51,43 @@ def _use_unicode(*args):
     Helper function for determining the output type of some string
     operations.
 
-    For an operation on two ndarrays, if at least one is unicode, the
-    result should be unicode.
+    For an operation on two ndarrays, if at least one is a dtype with a
+    scalar that subclasses str or bytes, the result should be that
+    dtype. Otherwise, if at least one is unicode, the result should be
+    unicode.
     """
     for x in args:
-        if (isinstance(x, str) or
-                issubclass(numpy.asarray(x).dtype.type, unicode_)):
+        scalar_type = numpy.asarray(x).dtype.type
+        if (isinstance(x, str) or issubclass(scalar_type, unicode_)):
             return unicode_
+        elif (issubclass(scalar_type, (str, bytes))):
+            return x.dtype
     return string_
 
-def _to_string_or_unicode_array(result):
+
+def _is_unicode(arr):
+    """Returns True if arr is a string or a string array with a dtype that
+    represents a unicode string, otherwise returns False.
+
+    """
+    scalar_type = numpy.asarray(arr).dtype.type
+    if isinstance(arr, str) or issubclass(scalar_type, (unicode_, str)):
+        return True
+    return False
+
+
+def _to_string_or_unicode_array(result, output_dtype_like=None):
     """
-    Helper function to cast a result back into a string or unicode array
-    if an object array must be used as an intermediary.
+    Helper function to cast a result back into an array
+    with the appropriate dtype if an object array must be used
+    as an intermediary.
     """
-    return numpy.asarray(result.tolist())
+    ret = numpy.asarray(result.tolist())
+    dtype = getattr(output_dtype_like, 'dtype', None)
+    if dtype is not None:
+        return ret.astype(type(dtype)(_get_num_chars(ret)))
```

Comment:
Hmmm, at least add a `copy=False` to the `astype` call here.  I guess the objects should usually be all bytes or strings and thus the array already be correct (typically).

We can drop the `.tolist()` also, unless I am very mistaken.  The reason for that smells like a bug I fixed 3 years ago.

For whatever you have, I guess the subclass info (and thus correct type) may get lost.  It should possibly be spelled as `asarray(result.tolist(), dtype=type(output_dtype_like.dtype))`, but we can't easily yet.

Reply:
It looks like I need to keep the `tolist()`, at least for `partition` to work correctly:

```
Traceback (most recent call last):
  File "/home/nathan/.pyenv/versions/3.10.8-debug/lib/python3.10/pdb.py", line 1726, in main
    pdb._runscript(mainpyfile)
  File "/home/nathan/.pyenv/versions/3.10.8-debug/lib/python3.10/pdb.py", line 1586, in _runscript
    self.run(statement)
  File "/home/nathan/.pyenv/versions/3.10.8-debug/lib/python3.10/bdb.py", line 597, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/home/nathan/Documents/numpy-experiments/setup-asciidtype.py", line 38, in <module>
    print(repr(np.char.partition(arr, "r")))
  File "<__array_function__ internals>", line 185, in partition
  File "/home/nathan/Documents/numpy/numpy/core/defchararray.py", line 1244, in partition
    return _to_string_or_unicode_array(
  File "/home/nathan/Documents/numpy/numpy/core/defchararray.py", line 90, in _to_string_or_unicode_array
    return ret.astype(type(dtype)(_get_num_chars(ret)), copy=False)
TypeError: Can only store ASCII text in a ASCIIDType array.
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> /home/nathan/Documents/numpy/numpy/core/defchararray.py(90)_to_string_or_unicode_array()
-> return ret.astype(type(dtype)(_get_num_chars(ret)), copy=False)
(Pdb) p ret
array([('this', '', ''), ('is', '', ''), ('an', '', ''),
       ('a', 'r', 'ray')], dtype=object)
(Pdb) p numpy.asarray(result.tolist())
array([['this', '', ''],
       ['is', '', ''],
       ['an', '', ''],
       ['a', 'r', 'ray']], dtype='<U4')
```

I don't think this is a problem for `np.str_` or `np.unicode_` so indeed we could get rid of `tolist()` for those, but for asciidtype we end up passing asciidtype's `setitem` a tuple of strings, which I would expect numpy to have already processed. I'm also a tiny bit nervous about breaking code using numpy in unexpected ways, so I think I'd prefer to just leave the `tolist()` as-is for now.

Reply:
Yeah, OK, lets keep it, it would a followup issue anyway really.  I suspect `partition` is just badly written here, since it should store the 3 objects into a an `(..., 3)` object array right away (and doesn't).  But probably that goes all the way into `_vec_string`, so...

Thanks for having a look!